File: scripts/migrate_schema_v3.py
#!/usr/bin/env python3

"""
Schema v3.0 Migration Script

Changes:
1. Reorder item fields: item_sequence → exercise_name → equipment_key → prescription → performed
2. Convert weight fields from simple values to {value, unit} structure

Usage: python3 scripts/migrate_schema_v3.py
"""

import json
import os
from pathlib import Path
from typing import Any, Dict, List

GOLDEN_SET_DIR = Path(__file__).parent.parent / 'data' / 'golden_set'

# Weight field patterns to transform
WEIGHT_FIELDS = [
    'actual_weight_kg',
    'target_weight_kg',
    'target_weight_kg_min',
    'target_weight_kg_max',
    'target_load',
    'actual_load'
]

# Statistics
stats = {
    'files_processed': 0,
    'items_reordered': 0,
    'weight_fields_converted': 0,
    'errors': []
}


def convert_weight(value: Any) -> Any:
    """Convert weight value to new structure"""
    if value is None:
        return None
    
    # Handle arrays (multiple sets)
    if isinstance(value, list):
        return [{'value': v, 'unit': 'kg'} for v in value]
    
    # Handle single values
    return {'value': value, 'unit': 'kg'}


def transform_weights(obj: Any) -> None:
    """Transform weight fields in an object"""
    if not isinstance(obj, dict):
        return
    
    # Handle min/max range pattern
    if 'target_weight_kg_min' in obj and 'target_weight_kg_max' in obj:
        obj['target_weight'] = {
            'value_min': obj['target_weight_kg_min'],
            'value_max': obj['target_weight_kg_max'],
            'unit': 'kg'
        }
        del obj['target_weight_kg_min']
        del obj['target_weight_kg_max']
        stats['weight_fields_converted'] += 2
    
    # Handle individual weight fields
    for field in WEIGHT_FIELDS:
        if field in obj and field not in ['target_weight_kg_min', 'target_weight_kg_max']:
            new_field = field.replace('_kg', '')
            obj[new_field] = convert_weight(obj[field])
            del obj[field]
            stats['weight_fields_converted'] += 1
    
    # Recursively process nested objects and arrays
    for key, value in list(obj.items()):
        if isinstance(value, list):
            for item in value:
                transform_weights(item)
        elif isinstance(value, dict):
            transform_weights(value)


def reorder_item_fields(item: Dict[str, Any]) -> Dict[str, Any]:
    """Reorder fields in an item object"""
    if not isinstance(item, dict):
        return item
    
    # Check if this is an item with exercise fields
    if not ('exercise_name' in item or 'exercises' in item or 'exercise_options' in item):
        return item
    
    # Create new dict with desired field order
    ordered_item = {}
    
    # 1. Sequence
    if 'item_sequence' in item:
        ordered_item['item_sequence'] = item['item_sequence']
    
    # 2. Identity fields (exercise_name, equipment_key)
    if 'exercise_name' in item:
        ordered_item['exercise_name'] = item['exercise_name']
    if 'equipment_key' in item:
        ordered_item['equipment_key'] = item['equipment_key']
    
    # 3. Prescription
    if 'prescription' in item:
        ordered_item['prescription'] = item['prescription']
    
    # 4. Performed
    if 'performed' in item:
        ordered_item['performed'] = item['performed']
    
    # 5. All other fields
    for key, value in item.items():
        if key not in ordered_item:
            ordered_item[key] = value
    
    stats['items_reordered'] += 1
    return ordered_item


def process_items(obj: Any) -> None:
    """Process all items in blocks recursively"""
    if not isinstance(obj, dict):
        return
    
    # If this is an items array, reorder each item
    if 'items' in obj and isinstance(obj['items'], list):
        obj['items'] = [reorder_item_fields(item) for item in obj['items']]
        
        # Process nested exercises/exercise_options
        for item in obj['items']:
            if 'exercises' in item and isinstance(item['exercises'], list):
                item['exercises'] = [reorder_item_fields(ex) for ex in item['exercises']]
            if 'exercise_options' in item and isinstance(item['exercise_options'], list):
                item['exercise_options'] = [reorder_item_fields(ex) for ex in item['exercise_options']]
    
    # Recursively process nested structures
    for key, value in obj.items():
        if isinstance(value, list):
            for item in value:
                process_items(item)
        elif isinstance(value, dict):
            process_items(value)


def process_file(filename: str) -> None:
    """Process a single JSON file"""
    filepath = GOLDEN_SET_DIR / filename
    
    try:
        print(f"Processing: {filename}")
        
        # Read and parse JSON
        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Step 1: Reorder fields
        process_items(data)
        
        # Step 2: Transform weights
        transform_weights(data)
        
        # Write back with pretty formatting
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
            f.write('\n')
        
        stats['files_processed'] += 1
        print(f"✓ {filename} updated")
        
    except Exception as error:
        stats['errors'].append({'file': filename, 'error': str(error)})
        print(f"✗ Error processing {filename}: {error}")


def main():
    """Main execution"""
    print('🚀 Starting Schema v3.0 Migration\n')
    print('Changes:')
    print('  1. Reorder item fields (exercise_name, equipment_key before prescription/performed)')
    print('  2. Convert weight fields to {value, unit} structure\n')
    
    # Get all JSON files
    files = sorted([f.name for f in GOLDEN_SET_DIR.glob('*.json')])
    
    print(f'Found {len(files)} JSON files\n')
    
    # Process each file
    for filename in files:
        process_file(filename)
    
    # Print summary
    print('\n' + '=' * 60)
    print('📊 Migration Summary')
    print('=' * 60)
    print(f"Files processed: {stats['files_processed']}/{len(files)}")
    print(f"Items reordered: {stats['items_reordered']}")
    print(f"Weight fields converted: {stats['weight_fields_converted']}")
    
    if stats['errors']:
        print(f"\n❌ Errors: {len(stats['errors'])}")
        for err in stats['errors']:
            print(f"  - {err['file']}: {err['error']}")
    else:
        print('\n✅ Schema v3.0 migration complete!')


if __name__ == '__main__':
    main()

File: scripts/migrate_schema_v3.js
#!/usr/bin/env node

/**
 * Schema v3.0 Migration Script
 * 
 * Changes:
 * 1. Reorder item fields: item_sequence → exercise_name → equipment_key → prescription → performed
 * 2. Convert weight fields from simple values to {value, unit} structure
 * 
 * Usage: node scripts/migrate_schema_v3.js
 */

const fs = require('fs');
const path = require('path');

const GOLDEN_SET_DIR = path.join(__dirname, '..', 'data', 'golden_set');

// Weight field patterns to transform
const WEIGHT_FIELDS = [
  'actual_weight_kg',
  'target_weight_kg',
  'target_weight_kg_min',
  'target_weight_kg_max',
  'target_load',
  'actual_load'
];

// Statistics
let stats = {
  filesProcessed: 0,
  itemsReordered: 0,
  weightFieldsConverted: 0,
  errors: []
};

/**
 * Convert weight value to new structure
 */
function convertWeight(value, field) {
  if (value === null || value === undefined) {
    return null;
  }

  // Handle arrays (multiple sets)
  if (Array.isArray(value)) {
    return value.map(v => ({ value: v, unit: 'kg' }));
  }

  // Handle single values
  return { value: value, unit: 'kg' };
}

/**
 * Transform weight fields in an object
 */
function transformWeights(obj) {
  if (!obj || typeof obj !== 'object') {
    return;
  }

  // Handle min/max range pattern
  if (obj.target_weight_kg_min !== undefined && obj.target_weight_kg_max !== undefined) {
    obj.target_weight = {
      value_min: obj.target_weight_kg_min,
      value_max: obj.target_weight_kg_max,
      unit: 'kg'
    };
    delete obj.target_weight_kg_min;
    delete obj.target_weight_kg_max;
    stats.weightFieldsConverted += 2;
  }

  // Handle individual weight fields
  WEIGHT_FIELDS.forEach(field => {
    if (obj[field] !== undefined) {
      // Determine new field name
      let newField = field.replace('_kg', '');
      
      // Skip if already converted as part of min/max
      if (field === 'target_weight_kg_min' || field === 'target_weight_kg_max') {
        return;
      }

      obj[newField] = convertWeight(obj[field], field);
      delete obj[field];
      stats.weightFieldsConverted++;
    }
  });

  // Recursively process nested objects and arrays
  Object.keys(obj).forEach(key => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach(item => transformWeights(item));
    } else if (typeof obj[key] === 'object' && obj[key] !== null) {
      transformWeights(obj[key]);
    }
  });
}

/**
 * Reorder fields in an item object
 */
function reorderItemFields(item) {
  if (!item || typeof item !== 'object') {
    return item;
  }

  // Check if this is an item with exercise_name (identity fields)
  if (!item.exercise_name && !item.exercises && !item.exercise_options) {
    return item;
  }

  // Create new object with desired field order
  const orderedItem = {};
  
  // 1. Sequence
  if (item.item_sequence !== undefined) {
    orderedItem.item_sequence = item.item_sequence;
  }

  // 2. Identity fields (exercise_name, equipment_key)
  if (item.exercise_name !== undefined) {
    orderedItem.exercise_name = item.exercise_name;
  }
  if (item.equipment_key !== undefined) {
    orderedItem.equipment_key = item.equipment_key;
  }

  // 3. Prescription
  if (item.prescription !== undefined) {
    orderedItem.prescription = item.prescription;
  }

  // 4. Performed
  if (item.performed !== undefined) {
    orderedItem.performed = item.performed;
  }

  // 5. All other fields (circuit_config, exercises, exercise_options, etc.)
  Object.keys(item).forEach(key => {
    if (!orderedItem.hasOwnProperty(key)) {
      orderedItem[key] = item[key];
    }
  });

  stats.itemsReordered++;
  return orderedItem;
}

/**
 * Process all items in blocks recursively
 */
function processItems(obj) {
  if (!obj || typeof obj !== 'object') {
    return;
  }

  // If this is an items array, reorder each item
  if (obj.items && Array.isArray(obj.items)) {
    obj.items = obj.items.map(item => {
      // Reorder top-level item
      let reorderedItem = reorderItemFields(item);
      
      // Process nested exercises/exercise_options
      if (reorderedItem.exercises && Array.isArray(reorderedItem.exercises)) {
        reorderedItem.exercises = reorderedItem.exercises.map(ex => reorderItemFields(ex));
      }
      if (reorderedItem.exercise_options && Array.isArray(reorderedItem.exercise_options)) {
        reorderedItem.exercise_options = reorderedItem.exercise_options.map(ex => reorderItemFields(ex));
      }
      
      return reorderedItem;
    });
  }

  // Recursively process nested structures
  Object.keys(obj).forEach(key => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach(item => processItems(item));
    } else if (typeof obj[key] === 'object' && obj[key] !== null) {
      processItems(obj[key]);
    }
  });
}

/**
 * Process a single JSON file
 */
function processFile(filename) {
  const filepath = path.join(GOLDEN_SET_DIR, filename);
  
  try {
    console.log(`Processing: ${filename}`);
    
    // Read and parse JSON
    const content = fs.readFileSync(filepath, 'utf8');
    const data = JSON.parse(content);
    
    // Step 1: Reorder fields
    processItems(data);
    
    // Step 2: Transform weights
    transformWeights(data);
    
    // Write back with pretty formatting
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2) + '\n', 'utf8');
    
    stats.filesProcessed++;
    console.log(`✓ ${filename} updated`);
    
  } catch (error) {
    stats.errors.push({ file: filename, error: error.message });
    console.error(`✗ Error processing ${filename}:`, error.message);
  }
}

/**
 * Main execution
 */
function main() {
  console.log('🚀 Starting Schema v3.0 Migration\n');
  console.log('Changes:');
  console.log('  1. Reorder item fields (exercise_name, equipment_key before prescription/performed)');
  console.log('  2. Convert weight fields to {value, unit} structure\n');
  
  // Get all JSON files
  const files = fs.readdirSync(GOLDEN_SET_DIR)
    .filter(f => f.endsWith('.json'))
    .sort();
  
  console.log(`Found ${files.length} JSON files\n`);
  
  // Process each file
  files.forEach(processFile);
  
  // Print summary
  console.log('\n' + '='.repeat(60));
  console.log('📊 Migration Summary');
  console.log('='.repeat(60));
  console.log(`Files processed: ${stats.filesProcessed}/${files.length}`);
  console.log(`Items reordered: ${stats.itemsReordered}`);
  console.log(`Weight fields converted: ${stats.weightFieldsConverted}`);
  
  if (stats.errors.length > 0) {
    console.log(`\n❌ Errors: ${stats.errors.length}`);
    stats.errors.forEach(err => {
      console.log(`  - ${err.file}: ${err.error}`);
    });
  } else {
    console.log('\n✅ Schema v3.0 migration complete!');
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { processFile, transformWeights, reorderItemFields };
